Feature/collect link metrics

mariomaz opened the pull request:
See "[TASK] Link metric collection for Ethernet links #390" for a detailed description of the task. 

Currently we only support collecting link metrics for Ethernet interfaces and it is done partially, because we need topology information to know which are the neighbors of the node. This is not available yet. 

Collecting metrics for WiFi interfaces requires adding support for standard commands in DWPAL. See "Send standard NL80211 commands using DWPAL #782" for more information.

For tests_flows.sh to pass, "Fix several problems with the Device Information TLV #825" should be merged first. 

I preferred to create a draft-PR so you can verify the design of the solution and give early feedback. Also to be particularly reviewed is the way MAC address of requested specific neighbor is compared against MAC address of the controller.

mariomaz committed:
agent: add link_metrics_collector interface and ethernet implementation

Define a structure to hold collected link metrics. This structure will
later be used to fill the TLVs of the Link Metric response message.

Use uint32_t instead of uint16_t for max throughput capacity and bit
rate fields to allow for bigger values.
Spec document uses uint16_t so a saturation to the maximum value is
required when copying values to the TLV.

Link metrics have to be collected differently for wired and wireless
links. Define an interface class for a link metrics collector. Derived
implementation classes will override pure virtual methods.

Create an implementation class for Ethernet links. This implementation
reads network interface stats using a Netlink socket and the
NETLINK_ROUTE protocol.

TODO in future tasks:
- Implement link metrics collector for WiFi links in ([TASK] Link
metric collection for WiFi links #792)
- Move all link metrics related classes to BPL ([TASK] Move link
metric related classes to BPL #910)
- Implement Ethernet link metrics collector using and extending Netlink
classes defined in "Feature/add support for nl80211 standard commands
to BWL #900"

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: update auto generated files

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: backhaul_manager: add handler for link metric query message

Add handler for LINK_METRIC_QUERY_MESSAGE message. On reception, obtain
requested link metrics, build LINK_METRIC_RESPONSE_MESSAGE and send it.

Currently only wired interfaces are supported. Obtaining link metrics
for wireless interfaces will be done in a future task.

Topology database is required to know which neighbor nodes are
connected to each interface. Since we do not have such database now, a
temporary solution has been implemented: a function that returns a list
with a single (neighbor, interface) pair for the controller. In the
future, once we have the topology database available, this function
will completed to return all links between interfaces and neighbors.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
controller: son_slave: get log messages ready for testing

Modify the strings written to log so test_flows can later check
which messages and TLVs have been received.

Fix some typos in log messages.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
test_flows: create test for link metrics query/response

Update test_flows.py to create test for LINK_METRIC_QUERY_MESSAGE
reception in agent.

Check also that controller receives LINK_METRIC_RESPONSE_MESSAGE.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: son_slave: remove handler for link metric query message

Link Metric Query message is handled now in backhaul manager thread
so remove its (empty) handler from son slave thread.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: add media type group

Media type group is the MSB of the media type field. It is not included
in any message as is. It is used internally to decide which mechanism
has to be used to compute the media type.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: fix a typo error

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: split link metric query TLV into two TLVs

The Link Metric Query TLV contains an optional field. However, TLVF
does not currently support optional fields.

As a workaround we create two different TLVs, one that includes the
optional field and another one which does not. When parsing received
message, the application will check the length of included TLV to know
which one is it.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: swap returned value of getNextTlvLength()

Method getNextTlvLength() returns an uit16_t with the length of next
TLV in message. However, the method is not considering the endianness
and returns the 2 bytes swapped.

Fix getNextTlvLength to consider endianness swapping bytes of returned
length value.

Swapping is only correct in the context of CmduMessageRx::parse().
Remove getNextTlvData and getNextTlvType(eTlvType) from CmduMessage.
Move getNextTlvType() and getNextTlvLength down into CmduMessageRx and
make them private.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: check length of received Link Metric Query TLV

Check the length of received TLV to know if optional field (MAC address
of neighbor device) is present or not and then create an instance of
either tlvLinkMetricQuerySpecificNeighbor or tlvLinkMetricQuery
respectively.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: use renamed Link Metric Query TLV

After having created different TLVs to cope with optional field in
tlvLinkMetricQuery and moved some inner enums out of their classes,
update test that was was using such TLV.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

arnout commented on agent/src/beerocks/slave/link_metrics/link_metrics.h:
```diff
@@ -0,0 +1,123 @@
+    /**
+     * The maximum MAC throughput of the link estimated at the transmitter and expressed in Mb/s.
+     */
+    uint16_t mac_throughput_capacity;
```
It is better to put the unit explicitly in the name, i.e. `mac_throughput_capacity_mbps`.

Also, is it a good idea to limit to 64Gbps? 10Gbps links are a reality in commercial switches, so 64Gbps is not that far off... The spec indeed puts it as a uint16_t, but I think it's better to use a generic unsigned or uint32_t here and do saturation when converting to the TLV.

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/link_metrics/link_metrics.h:
```diff
@@ -0,0 +1,123 @@
+     * the PHY rate estimated at the transmitter of the link expressed in Mb/s; otherwise, it is
+     * set to 0xFFFF.
+     */
+    uint16_t phy_rate;
```
Same question about the uint16_t.

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/link_metrics/link_metrics.h:
```diff
@@ -0,0 +1,123 @@
+     * @return True on success and false otherwise.
+     */
+    virtual bool get_link_metrics(const std::string &local_interface_name,
+                                  const sMacAddr &neighbor_interface_address,
```
I think in practice it will never be possible to distinguish between different neighbours that are connected on the same interface. This is a serious shortcoming of the 1905.1 spec, but not much we can do about it... @mariomaz do you think you can ask your colleagues who are more familiar with 1905.1 how you are supposed to deal with interfaces where there are multiple neighbours, or none?

Anyway, my point is: I think in practice we will only use the local_interface_name.

(nitpick)

mariomaz replied:
I agree with you.
I've checked our 1905.1 implementation at this point and found this TODO which also confirms your opinion:
```c
            //   TODO: In Linux there is no easy way to obtain this
            //   information. We will just report the amount of *total* packets
            //   transmitted from our local interface, no matter the
            //   destination.
            //   This information will only be correct when the local interface
            //   is connected to one single remote interface... however we
            //   better report this than nothing at all.
```
However, be aware that this is an interface which will be implemented by different collectors, not only the 802.3 collector. In the 802.11 collector it is possible to distinguish between different neighbors and thus we do need the neighbor_interface_address parameter.

arnout commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+    rtnl_msg.msg_iov     = &io;
+    rtnl_msg.msg_iovlen  = 1;
+    rtnl_msg.msg_name    = &kernel;
+    rtnl_msg.msg_namelen = sizeof(kernel);
```
Is there any reason to open-code all this, rather than using libnl3 genl? In fact, we could move base_wlan_hal_nl80211::send_nl80211_msg to network_utils and use that function. The nl80211 part of it is only hardcoded in the genl_ctrl_resolve call.

Anyway, something that can be done later.

mariomaz replied:
I agree with you. 

This is something that could probably be done using changes in "Feature/add support for nl80211 standard commands to BWL #900". However, be aware that such changes have been made into BWL and not in network_utils. 

Anyway, I agree with you that this is a refactoring that could be done later (I think it is not going to be so easy and quick as might seem at first glance).

tomereli replied:
Why not indeed move `base_wlan_hal_nl80211::send_nl80211_msg` to BCL? just that function, shouldn't be too hard even as a followup commit in this PR. I really prefer that we don't duplicate code when we really don't need to. Still, as you say, it can be done later, so (nitpoick)

mariomaz replied:
Method base_wlan_hal_nl80211::send_nl80211_msg() is going to disappear after we had approved the design and merged "Feature/add support for nl80211 standard commands to BWL #900" and done a refactoring of the NL80211 version of the BWL. 

The work to be done is then to implement the get_link_metrics() method using the new classes defined in #900.

arnout replied:
Regardless of the new classes, the question remains why you would do all the parsing manually rather than use one of the rtnl_link_get_ functions to get all info in a nice struct.

arnout commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+            /**
+             * This message contains the stats for the interface we are interested in
+             */
+            if (0 == strcmp(local_interface_name.c_str(), (char *)RTA_DATA(attribute))) {
```
I believe there is a way to specify this already in the GET message, so instead of getting all interfaces, you get just one.

You would need to convert it to ifindex first though, IIRC.

(nitpick).

mariomaz replied:
Since this is a nitpick comment, the change could probably be done as part of the previous refactoring.

tomereli replied:
strcmp is baned C function if I'm not mistaken.

tomereli replied:
So it appears I was mistaken and apparently the list of banned C functions by Intel SDL is listed in github:
https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
We should add this to our CONTRIBUTING.md, perhaps @vitalii-komisarenko @abelog or @VladyslavTupikin can do that as good quick first PR.

mariomaz replied:
Although strcmp is not in the list of banned C functions, I've replaced it with std::strncmp() C++ function which is safer.

arnout commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+    struct rtgenmsg gen;
+};
+
+static bool get_link_metrics(const struct nlmsghdr *h, const std::string &local_interface_name,
```
I think this function should get a slightly more specific name.

mariomaz replied:
I'm not sure about what you mean by "more specific". If you please suggest a better name I will change it. However, be aware that this code will be refactored and this function will disappear.

tomereli replied:
Can you elaborate on the fact that this will disappear? You mean it will be part of the discovery service? In that case it is worth to mention that in the commit message.

mariomaz replied:
This function will disappear once we do a refactor of this class to use Netlink classes defined in #900 to get link metrics.

arnout commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+                 * things even simpler, we will set the MAC throughput capacity with the same
+                 * value as the physical data rate, as if there was no overhead at all.
+                 */
+                link_metrics.transmitter.mac_throughput_capacity = phy_rate;
```
For Ethernet, we do know the overhead, or at least we can estimate it: 14 bytes header and 4 bytes CRC plus 12 bytes of interframe gap on a 1500 byte payload. So 1500/1530. But again, can be done later.

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+                        phy_rate = 100;
+                    } else if (SPEED_1000 == speed) {
+                        phy_rate = 1000;
+                    }
```
Here, there really is no reason at all to limit to specific speeds. We can simply use the speed value directly. The only thing we have to do is to saturate at 64Gbps, but that's something which we should do while converting to TLV (as mentioned below).

In other words, I think we can pass phy_rate directly into linux_iface_get_speed.

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+                    }
+                }
+
+                link_metrics.transmitter.packet_errors       = stats->tx_errors;
```
I believe we have to do this right...

All the stats fields are uint32_t and wrap. tx_errors probably won't wrap, but the number of rx/tx packets will wrap if the device stays up for a couple of days.

Therefore, we have to keep the previously measured value and subtract it (wrapping subtraction, I'm not entirely sure what the standard C++ way of doing that is).

While we're at it, we should probably also keep the timestamp. In 1905.1 the time of the measurement isn't specified (which renders it completely useless), but in R2 there is a timestamp included. And to make the metrics somewhat useful even in R1, we should probably divide by the sampling time so the numbers are at least comparable with each other.

So, I think it would be better to keep an object for the link metrics for each interface.

That also means you can keep the netlink fd open. Minor optimisation.

mariomaz replied:
Sorry but I cannot see the timestamp field you mentioned. Maybe I'm not looking at the right document.  Can you please clarify which are those R1 and R2 docs?
I agree with you that if we cannot include the time measurement window along with the counter, then it is useless to provide the number of packets/errors **during** such measurement period. 

I think however that, even if we could provide timestamp or elapsed time since last measure, we have no way to know if the number of packets we read with the netlink socket have wrapped (and how many times) since last call to get link metrics. Although such wrapping is quite improbable to go unnoticed, given the frequency with which calls to get link metrics will be probably issued, we cannot guarantee that wrapping won't happen. Keeping timestamps is useless if they are not provided by the kernel.

arnout replied:
> Sorry but I cannot see the timestamp field you mentioned.

I should have written "we should keep *a* timestamp".

> we have no way to know if the number of packets we read with the netlink socket have wrapped (and how many times) since last call to get link metrics.

Stupid me again, I thought these were 64-bit numbers, but they're only 64-bit if you get them through netlink, not if you get them from IFLA_STATS64. Which apparently requires special handling because they're unaligned...

Which brings me to the question: why don't we use rtnl for this? E.g. https://gist.github.com/beejeebus/cc4fb07472cf5a0afd41 rtnl_link takes care of the 64-bit handling and fallback to 32-bit if the 64-bit stats are not available.

arnout replied:
Just to be clear: since the code is there now anyway, I would merge it as is. But when later we refactor with #900, I think we should also look into using rtnl_link instead of doing all the parsing manually. However, before going into that, we should check if the libnl-tiny used in openwrt also has it.

mariomaz replied:
Read in https://wireless.wiki.kernel.org/en/developers/documentation/nl80211

> libnl tiny
> OpenWrt folks created a tiny version of libnl based on a git snapshot, which only contains genl, not rtnetlink or any of the netfilter stuff, and compiles down to less than 30k in binary size. You can find it here:
> https://git.openwrt.org/?p=openwrt/openwrt.git;a=tree;f=package/libs/libnl-tiny;hb=HEAD

arnout commented on framework/tlvf/yaml/tlvf/ieee_1905_1/eLinkMetricNeighborType.yaml:
```diff
@@ -0,0 +1,10 @@
+#
+---
+_namespace: ieee1905_1
+
+eLinkMetricNeighborType:
+  _type: enum
+  _enum_storage: uint8_t
+  ALL_NEIGHBORS: 0x00
+  SPECIFIC_NEIGHBOR: 0x01
```
Moving these to a separate file wasn't really needed, but OK.

mariomaz replied:
If you declare eLinkMetricNeighborType in the same YAML file as the two TLV classes that use it then, as TLVF defines the eLinkMetricNeighborType **inside** one of the TLV C++ classes, you get a compilation error because it's undefined for the other TLV class.

Maybe there is some other workaround for this but I found moving eLinkMetricNeighborType to a different file the easiest.

arnout replied:
I didn't realize that. OK.

arnout commented on framework/tlvf/yaml/tlvf/ieee_1905_1/tlvLinkMetricQuery.yaml:
```diff
@@ -2,29 +2,30 @@
     _value_const: TLV_LINK_METRIC_QUERY
   length: uint16_t
-  neighbor_type: eNeighborType
+  neighbor_type: eLinkMetricNeighborType
```
You could make this
```
  neighbor_type:
    _type: eLinkMetricNeighborType
    _value_const: ALL_NEIGHBORS
```

Note that you can't do that for the other one.

mariomaz replied:
Done

arnout commented on framework/tlvf/yaml/tlvf/ieee_1905_1/tlvLinkMetricQuery.yaml:
```diff
@@ -2,29 +2,30 @@
+  neighbor_type: eLinkMetricNeighborType
+  link_metrics_type: eLinkMetricsType
+
+tlvLinkMetricQuerySpecificNeighbor:
```
This is not named correctly. It will also be used for the All Neighbors case for devices that do include a 00:00:00:00:00:00 MAC address.

mariomaz replied:
Renamed tlvLinkMetricQuerySpecificNeighbor with its original name tlvLinkMetricQuery because it's suitable for both use cases: all-neighbors and specific-neighbor

arnout commented on framework/tlvf/yaml/tlvf/ieee_1905_1/tlvLinkMetricQuery.yaml:
```diff
@@ -2,29 +2,30 @@
+# optional field. Application must then check the length of received message to know if optional 
+# field is present or not and then create an instance of either of these classes.
+
+tlvLinkMetricQueryAllNeighbors:
```
Since the tlvLinkMetricQuery class is removed here, we also have to remove all instances of it in the same commit. It is used in the parse function in CmduMessageRx and in tlvf_test.

To split up the commits a bit more, you could first simply add the two new ones but keep the old one around, then update the parse function and tlvf_test, and then remove the old one.

mariomaz replied:
With previous renaming, tlvLinkMetricQuery is being used back again.

arnout reviewed (commented):
Reviewed up to "split link metric query TLV into two TLVs".

arnout commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -74,7 +74,30 @@ std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv()
+        } else if (min_tlv_length == tlv_length) {
+            return msg.addClass<tlvLinkMetricQueryAllNeighbors>();
+        } else {
+            return msg.addClass<tlvUnknown>();
```
Maybe there should be a warning here.

mariomaz replied:
Since now we have tlvLinkMetricQuery back again, I've changed it to use tlvLinkMetricQuery when the length of the TLV does not match the length of tlvLinkMetricQueryAllNeighbors.
tlvUnknown is now used by the default switch clause only.

arnout commented on framework/tlvf/src/src/CmduMessage.cpp:
```diff
@@ -32,7 +32,11 @@ uint16_t CmduMessage::getNextTlvLength() const
-    return tlv->length;
+
+    uint16_t tlv_length = tlv->length;
+    swap_16(tlv_length);
```
This swapping is only correct in the context where you call it, i.e. in CmduMessageRx::parse(). In the one case where it is currently used, in a debug message in tlvf_test, it actually should not be swapped, since it's already swapped by the parse() that precedes it.

Therefore, I propose to move this function down into CmduMessageRx first and make it private.

And while you're at it, remove getNextTlvData and getNextTlvType(eTlvType), and move getNextTlvType() down into CmduMessageRx as well. But maybe that's a bit too much for this PR :-).

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2069,6 +2267,186 @@ bool backhaul_manager::handle_1905_higher_layer_data_message(ieee1905_1::CmduMes
+        /**
+    	   * If tlvLinkMetricQuerySpecificNeighbor has been included in message, we will be
+    	   * permissive enough to allow it specify ALL_NEIGHBORS and if so, then we will just
+    	   * ignore the field containing the MAC address of neighbor.
```
:1st_place_medal:

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2069,6 +2267,186 @@ bool backhaul_manager::handle_1905_higher_layer_data_message(ieee1905_1::CmduMes
+    if (ieee1905_1::eLinkMetricNeighborType::SPECIFIC_NEIGHBOR == neighbor_type) {
+
+        // TODO: check if specified neighbor is invalid (topology database is required for that)
+    }
```
Nice preparation for the future! I like it!

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2069,6 +2267,186 @@ bool backhaul_manager::handle_1905_higher_layer_data_message(ieee1905_1::CmduMes
+        }
+
+        tlvLinkMetricResultCode->value() = ieee1905_1::tlvLinkMetricResultCode::INVALID_NEIGHBOR;
+    } else if (ieee1905_1::eLinkMetricNeighborType::SPECIFIC_NEIGHBOR == neighbor_type) {
```
(nitpick) I think it would be nicer to have a boolean "specific_neighbor" that is set above where we parse the two message types, so we can simply do `if (specific_neighbor)` here.

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2069,6 +2267,186 @@ bool backhaul_manager::handle_1905_higher_layer_data_message(ieee1905_1::CmduMes
+         * Obtain link metrics and add a Transmitter Link Metric TLV or a Receiver Link Metric TLV
+         * or both to response message.
+         */
+        if (!add_link_metrics(local_interface_name, media_type, reporter_al_mac, neighbor_al_mac,
```
I think we have to do it differently than the standard here. I don't think we'll ever support more than one neighbor on an interface. So somehow there's a one-to-one mapping between neighbors and interfaces.

So, I think we should get from the topology manager a list of (neighbor, interface) pairs, and then call add_link_metrics on it. The two al_mac addresses are there only to copy into the TLVs. In the specific neighbor case, the list is filtered to find the single instance that was requested - which also covers the invalid neighbor case.

I believe this approach will simplify things somewhat.

Concretely, since we don't have a topology manager yet, I would add a function that returns a list with a single (neighbor, interface) pair for the controller (which is BTW incorrect for the second repeater if we have a chain topology, because then the controller is not a neighbor). So that replaces the logic above.

mariomaz replied:
Very good point. 
I've included your suggestions to simplify code. However, one-to-many mapping between interfaces and neighbors is possible in WiFi so I've kept that possibility also for Ethernet (because it's simpler to code, doesn't hurt and who knows if someday someone connects more than one IEEE 1905 device to a switch).

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2069,6 +2267,186 @@ bool backhaul_manager::handle_1905_higher_layer_data_message(ieee1905_1::CmduMes
+             * Compute particular Ethernet media type
+             */
+            if (!get_media_type(local_interface_name, ieee1905_1::eMediaTypeGroup::IEEE_802_3,
+                                media_type)) {
```
I don't think it makes sense to do get_media_type at this level, I'd rather do it inside a dispatcher function in the link metrics collector.

Or better yet, for each interface I would like to have an actual object instead of just a name, and that object would have a reference to the collector (and the media type). But maybe that's going a bit too far for this PR...

mariomaz replied:
get_media_type() is an expensive function in terms of performance. If we take it out from where it is now, then we'll end up calling it more than once for the same interface.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -74,6 +77,198 @@ namespace beerocks {
+            }
+            auto interfacePairInfo                      = std::get<1>(interface_pair_info);
+            interfacePairInfo.rc_interface_mac          = local_interface_mac;
+            interfacePairInfo.neighbor_interface_mac    = neighbor_al_mac;
```
Hm... This is not correct... We actually have to get this from the topology discovery message, which will give us the neighbor interface MAC.

So for now, add a TODO.

mariomaz replied:
Done

mariomaz commented:
I've also replaced strcmp() banned function with std::strncmp()

tomereli commented on agent/src/beerocks/slave/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ set(MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
 
 file(GLOB beerocks_agent_sources 
     ${MODULE_PATH}/backhaul_manager/*.c*
+    ${MODULE_PATH}/link_metrics/*.c*
```
This is classic BPL implementation, so I would prefer it in the BPL.
But since you say below that this is a temporary solution, and this code should be refactored pretty soon so its part of the discovery service - (nitpick)

However, please add a comment here that this should move to the discovery agent and add a link to the github relevant task.

mariomaz replied:
Done.
I've created "[TASK] Move link metric related classes to BPL #910"
However, link metric collection is not meant to be part of the discovery service but it requires the topology database that will be populated by the discovery service.

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+    bool result = false;
+
+    struct ifinfomsg *iface = static_cast<ifinfomsg *>(NLMSG_DATA(h));
+    int length              = h->nlmsg_len - NLMSG_LENGTH(sizeof(*iface));
```
Any reason not to use `size_t`? Can `length` be negative?

mariomaz replied:
I've replaced strcmp with std::strncmp.
When I say that get_link_metrics() will disappear I mean that it's going to be refactored and coded in a different way.
I've changed the type of `length` from `int` to `size_t` and checked the values before subtracting.

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+     * Loop over all attributes of the NEWLINK message
+     */
+    for (struct rtattr *attribute = IFLA_RTA(iface); RTA_OK(attribute, length);
+         attribute                = RTA_NEXT(attribute, length)) {
```
Extra space

mariomaz replied:
clang-format.sh is to blame

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+    /**
+     * Netlink socket is ready for use, prepare and send request
+     */
+    memset(&rtnl_msg, 0, sizeof(rtnl_msg));
```
Although `memset` is not banned, it is recommended to use `memset_s` or an equivalent c++ STD function such as `std::fill`. See https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions

mariomaz replied:
I've set the initial value of structs to zero on declaration and removed all calls to `memset`. 
Looks like `memset_s` is not available.

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,246 @@
+    return result;
+}
+
+static bool get_link_metrics(const std::string &local_interface_name, sLinkMetrics &link_metrics)
```
I see now this is the 4th function overload with the same `get_link_metrics` name. This is fine, but can we at least add doxygen documentation to each function?

mariomaz replied:
Link metrics cannot be obtained with `base_wlan_hal_nl80211::send_nl80211_msg` as is. See https://github.com/prplfoundation/prplMesh/pull/831#discussion_r386899613

When you say BCL, I guess you mean BPL. Be aware that in #900 I've implemented the classes to send NL80211 commands and all Netlink related stuff in BWL. First of all we should decide  where those classes should be. 

I've documented the `get_link_metrics` functions.

tomereli reviewed (commented):
Partial review only for now, so just comments

mariomaz committed:
agent: backhaul_manager: refactor duplicated code

Now that we have a get_media_type() function, use it wherever it was
previously being computed a MediaType field.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
test_flows: remove emulated response from combined inframetrics test

Remove agent's emulated response because now the Link Metric query
message is processed by agent. Also, as the message handler is now in
the backhaul_manager_thread, the log file to look for messages has to
be changed.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

arnout commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -51,6 +51,26 @@
+uint16_t CmduMessageRx::getNextTlvLength() const
+{
+    if (!getCmduHeader())
+        return -1;
```
This would need an explicit cast to uint16_t IMO.

Or maybe even better: make the return value int.

mariomaz replied:
I've changed it to return UINT16_MAX

arnout commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -96,7 +96,27 @@ std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv()
     case (8): {
-        return msg.addClass<tlvLinkMetricQuery>();
+        /**
+         * The IEEE 1905.1 standard says about the Link Metric Query TLV and the neighbor type
+         * octet that "If the value is 0, then the EUI48 field is not present; if the value is 1,
+         * then the EUI-48 field shall be present."
+         *
+         * However, optional fields are not currently supported by TLVF.
+         *
+         * As a workaround, instead of defining a tlvLinkMetricQuery TLV with an optional field,
+         * we have defined two different TLVs, one with the optional field and the other one
+         * without it. Application must then check the length of received TLV to know if optional
+         * field (MAC address of neighbor device) is present or not and then create an instance of
+         * either tlvLinkMetricQuery or tlvLinkMetricQueryAllNeighbors respectively.
```
I don't know if I said it before, but this is really an excellent explanation.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2142,6 +2364,154 @@ bool backhaul_manager::handle_1905_higher_layer_data_message(ieee1905_1::CmduMes
+        link_metrics_type = tlvLinkMetricQueryAllNeighbors->link_metrics_type();
+    }
+
+    /**
+     * Set alias flag to true if link metrics for a specific neighbor have been requested
+     */
+    bool specific_neighbor =
+        ieee1905_1::eLinkMetricNeighborType::SPECIFIC_NEIGHBOR == neighbor_type;
```
(nitpick) I would set this directly in the condition above instead of saving neighbor_type.

arnout commented on tests/test_flows.sh:
```diff
@@ -200,6 +201,23 @@ test_client_capability_query() {
+    dbg "Confirming link metric response has been received on controller"
+    check_log ${GATEWAY} controller "Received LINK_METRIC_RESPONSE_MESSAGE"
+    check_log ${GATEWAY} controller "Received TLV_TRANSMITTER_LINK_METRIC"
+    check_log ${GATEWAY} controller "Received TLV_RECEIVER_LINK_METRIC"
```
This should of course be converted to Python.

mariomaz replied:
Done

arnout reviewed (approved):
Except for doing the test in python instead of shell, all my comments are nitpicky.

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+                 * payload. So 1500/1538.
+                 * (see https://en.wikipedia.org/wiki/Ethernet_frame)
+                 */
+                const uint32_t layer2_payload_size = 1500;
```
Why not define as `float` and get rid of the casting below?
```suggestion
                const float layer2_payload_size = 1500;
```
(nitpick)

mariomaz replied:
Done

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+                const uint32_t layer2_payload_size = 1500;
+                const uint32_t layer1_total_size   = 1538;
+                link_metrics.transmitter.mac_throughput_capacity_mbps =
+                    phy_rate_mbps * ((float)layer2_payload_size / (float)layer1_total_size);
```
C style casting should be avoided.

mariomaz replied:
Done

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
```
It took me some time to understand that this is only for ethernet links. I guess I should know by heart that 802.3 is Ethernet, but consider changing the file name to `eth_link_metrics_collector.cpp` (nitpick)

tomereli commented on agent/src/beerocks/slave/link_metrics/link_metrics.h:
```diff
@@ -0,0 +1,123 @@
+     * If the media type of the link is IEEE 802.11, then this value is the estimated RSSI in dB at
+     * the receive side of the link expressed in dB; otherwise, it is set to 0xFF.
+     */
+    uint8_t rssi = 0;
```
So why not set it to UINT8_MAX by default instead of 0? Then you don't need to set it in the ethernet link collector implementation.

mariomaz replied:
This struct is going to be used in the WiFi link collector implementation too. UINT8_MAX is not a default value, but the value to use for ethernet links (according to standard)

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+
+                link_metrics.receiver.packet_errors    = stats->rx_errors;
+                link_metrics.receiver.packets_received = stats->rx_packets;
+                link_metrics.receiver.rssi             = UINT8_MAX;
```
This should be the default in the struct so not needed to set here (nitpick)

mariomaz replied:
RSSI field must be set to UINT8_MAX in Ethernet metrics only. It's not the default value

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+{
+    bool result = false;
+
+    struct sockaddr_nl kernel; /* the remote (kernel space) side of the communication */
```
I think `socket` or `sock` is a more appropriate name here.
```suggestion
    struct sockaddr_nl sock; /* the remote (kernel space) side of the communication */
```
(nitpick)

mariomaz replied:
Done

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+     * process (no multicasting).
+     */
+    int fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+    if (fd < 0) {
```
Get rid of unnecessary indentation:
```suggestion
    if (fd < 0) {
        LOG(ERROR) << "Failed creating Netlink socket: " << strerror(errno);
        return false;
    }
```

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+ */
+static bool get_link_metrics(const std::string &local_interface_name, sLinkMetrics &link_metrics)
+{
+    bool result = false;
```
Not needed, just return true of false where needed.

mariomaz replied:
Single vs. Multiple exit point, that's the question.
Have you agreed on which one to use or is it up to the programmer? I haven't seen anything about this topic in `CODINGSTYLE.md`

tomereli replied:
I think in general we favour multiple exit points, it keeps the code indentation shorter and is what we have in most of prplmesh. I will update the `CODYINGSTULE.md` if agreed - @fvbogaert you have a chance for another poll :)

arnout replied:
We use multiple exit points everywhere. Because normally, all cleanup is automatic since we use smart pointers etc.

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+    int fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+    if (fd < 0) {
+        LOG(ERROR) << "Failed creating Netlink socket: " << strerror(errno);
+    } else {
```
```suggestion
    /**
         * Get link metrics using Netlink socket
         */
        result = get_link_metrics(fd, local_interface_name, link_metrics);

        /**
         * Clean up and finish properly
         */
        close(fd);
        return true;
```

mariomaz replied:
`result` from `get_link_metrics(fd, ...)` is being ignored with your proposal

mariomaz replied:
If we opt for multiple exit point we must call `close` twice.

tomereli replied:
If we're going for single exit point then we should go all the way and do it like in the kernel, with `goto`. @fvbogaert should we use `goto`? Another poll..

arnout replied:
Yes, that's my preferred pattern if there is any cleanup to do. However, we really should use automatic cleanup everywhere. So the better approach would be to wrap the file descriptor so it gets closed automatically by the destructor.

tomereli commented on agent/src/beerocks/slave/link_metrics/ieee802_3_link_metrics_collector.cpp:
```diff
@@ -0,0 +1,279 @@
+                                                        const sMacAddr &neighbor_interface_address,
+                                                        sLinkMetrics &link_metrics)
+{
+    (void)neighbor_interface_address;
```
We discussed this before - we should add an `UNUSED` macro instead (nitpick)

mariomaz replied:
Fixed. 
However, I've used the C++11 attribute syntax [[gnu::unused]] instead of __attribute__((unused))
This way we do not have to define the UNUSED in some common header file and include such header.

arnout replied:
Would be nice to add that to CODING_STYLE.md.

tomereli commented on framework/tlvf/AutoGenerated/include/tlvf/ieee_1905_1/eLinkMetricNeighborType.h:
```diff
@@ -0,0 +1,30 @@
+///////////////////////////////////////
+// AUTO GENERATED FILE - DO NOT EDIT //
+///////////////////////////////////////
+
```
This commit is misplaced as it seems - the previous commit did not change tlvf.py..

mariomaz replied:
This is not the first time I see this. Looks like github is messing the commits but they are actually ok. 
I've cloned the repository in another directory and double-checked that the order of these commits is correct.

tomereli replied:
@morantr @CoralMalachi I think you found a way to solve this in github?

tomereli commented on framework/tlvf/AutoGenerated/include/tlvf/ieee_1905_1/eMediaType.h:
```diff
@@ -36,6 +36,14 @@ enum eMediaType: uint16_t {
 };
 
+enum eMediaTypeGroup: uint8_t {
+    IEEE_802_3 = 0x0,
```
Adding comments (possible in `yaml` file) is welcome here

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -78,6 +81,222 @@ namespace beerocks {
+{
+    // TODO: Currently, only Ethernet is supported
+    // When dealing with WiFi interfaces, use NL80211 command to know if media type group is WiFi
+    media_type_group = ieee1905_1::eMediaTypeGroup::IEEE_802_3;
```
Which NL80211 command should be used to get the media type? Can you add this to the comment and perhaps a link to the task which will add it?

mariomaz replied:
Done

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -78,6 +81,222 @@ namespace beerocks {
+            result = true;
+        }
+    } else if (ieee1905_1::eMediaTypeGroup::IEEE_802_11 == media_type_group) {
+        // TODO: Not supported yet
```
Can we add LOG prints for these cases?

mariomaz replied:
Done

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -78,6 +81,222 @@ namespace beerocks {
+    if (0 == media_type_msb) {
+        collector = std::make_unique<ieee802_3_link_metrics_collector>();
+    } else if (1 == media_type_msb) {
+        // TODO: Create a collector for wireless interfaces
```
Should print LOG(ERROR) in this case for now.

mariomaz replied:
Done

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -78,6 +81,222 @@ namespace beerocks {
+        collector = std::make_unique<ieee802_3_link_metrics_collector>();
+    } else if (1 == media_type_msb) {
+        // TODO: Create a collector for wireless interfaces
+        //collector = std::make_unique<ieee802_11_link_metrics_collector>();
```
Why not create the collector here instead of putting dead code? You can just add a stub implementation for the 802_11 collector. (nitpick)

tomereli replied:
Now I see you dereference the collector below - we can just define the collector as nullptr, add a log print and a TODO.

mariomaz replied:
Done (The default initialization for unique_ptr points to nothing)

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -78,6 +81,222 @@ namespace beerocks {
+    /**
+     * Get link metrics information
+     */
+    if (collector &&
```
I think it is better to split to 2 if conditions (nitpick)

mariomaz replied:
Some static code analyzers would complain in such case.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -78,6 +81,222 @@ namespace beerocks {
+        network_utils::iface_info iface_info;
+        if (network_utils::get_iface_info(iface_info, local_interface_name) != 0) {
+            LOG(ERROR) << "Failed reading interface info for: " << local_interface_name;
+            return (false);
```
```suggestion
            return false;
```

mariomaz replied:
Done

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -78,6 +81,222 @@ namespace beerocks {
+         * Get local interface information to obtain its MAC address.
+         */
+        network_utils::iface_info iface_info;
+        if (network_utils::get_iface_info(iface_info, local_interface_name) != 0) {
```
You only use the mac address, so better to use:
```suggestion
        std::string mac;
        if (!network_utils::linux_iface_get_mac(local_interface_name, mac)) {
```

mariomaz replied:
Done

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2265,6 +2487,154 @@ bool backhaul_manager::handle_1905_higher_layer_data_message(ieee1905_1::CmduMes
+     * tlvLinkMetricQueryAllNeighbors without it. Application must check which of both TLVs has
+     * been received inside the message.
+     */
+    std::shared_ptr<ieee1905_1::tlvLinkMetricQueryAllNeighbors> tlvLinkMetricQueryAllNeighbors;
```
Perhaps we can handle each option in separate functions? (nitpick)

mariomaz replied:
Most of the code is common to both options. Separated functions would contain lots of duplicated code so I prefer to leave it as it is now.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2210,21 +2210,26 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
+     * Add a LocalInterfaceInfo field for the wired interface, if any.
+     */
+    std::string local_interface_name = m_sConfig.wire_iface;
+    // TODO: check if interface is UP?
```
Why not check it then?
For example:
```
 // If a wired (WAN) interface was provided, try it first, check if the interface is UP
            wan_monitor::ELinkState wired_link_state = wan_monitor::ELinkState::eInvalid;
            if (!m_sConfig.wire_iface.empty()) {
                wired_link_state = wan_mon.initialize(m_sConfig.wire_iface);
                // Failure might be due to insufficient permissions, datailed error message is being
                // printed inside.
                if (wired_link_state == wan_monitor::ELinkState::eInvalid) {
                    LOG(WARNING) << "wan_mon.initialize() failed, skip wired link establishment";
                }
            }
            if ((wired_link_state == wan_monitor::ELinkState::eUp) &&
                (selected_backhaul.empty() || selected_backhaul == DEV_SET_ETH)) {
```

mariomaz replied:
Done. 
However, I've used `network_utils::linux_iface_is_up_and_running()` to check interface status because it's simpler than your proposal and has no side effects.

tomereli reviewed (approved):
I have only nitpicks - this is an amazing work @mariomaz - well done!

mariomaz reviewed (commented):
Sorry, I pressed the wrong button
